### PR TITLE
buffer data structure for tee/fork

### PR DESCRIPTION
**Pull request by sithmel**
This should solve the issue https://github.com/sithmel/iter-tools/issues/156 and provides a base to develop *fork* easily.

I have created an intermediate abstraction called MessageQueue. I hate the name and I couldn't come up with a better metaphor. Apart that, I am pretty happy of how it works.

Comments welcome!

### CODE REVIEW

**Comment by sithmel** [comment]
When used as a queue, messageQueue performance are equivalent to dequeue (slightly better actually). So it may be worth dropping dequeue as a dependency.

```js
module.exports['dequeue 100000 items'] = function () {
  const q = new Dequeue()
  for (const n of range(10000)) {
    q.push(n)
  }
  while (q.length) {
    q.shift()
  }
}

module.exports['messageQueue 100000 items'] = function () {
  const q = new MessageQueue()
  for (const n of range(10000)) {
    q.add(n)
  }
  while (!q.isExhausted()) {
    q.get()
  }
}
```
```
  Dequeue against MessageQueue
    ✓  dequeue 10000 items x 2,558 ops/sec ±0.45% (92 runs sampled)
    ✓  messageQueue 10000 items x 2,708 ops/sec ±0.27% (94 runs sampled)

    Result: messageQueue 10000 items is 5.87% faster than dequeue 10000 items.
```

**Comment by conartist6** [comment]
Actually I needed to write my own queue for my optimal fork implementation because it needs to be able to swap implementations. I'll post the code when I'm done, probably tonight (California time)

**Comment by conartist6** [comment]
Mine should be more efficient than dequeue as well since it is singly linked.

**Comment by sithmel** [comment]
I have applied the new data structure everywhere and I got rid of dequeue.

I have now implemented 2 different abstractions, based on the same code:

### queue
Good old queue (single linked list) implementing:
* shift
* push
* isEmpty

All in O(1)

## exchange
I have borrowed the name from messaging protocol libraries (https://www.rabbitmq.com/tutorials/amqp-concepts.html#exchange-fanout).
The *exchange* can:
* create a consumer with **spawnConsumer**
* **push** an item in the queue
* use **noMoreConsumers** to declare that no more consumers will be created and therefore consumed items can be garbage collected

A *consumer* implements:
* shift
* isEmpty

So basically, you push items in the exchange and read them with the consumer.
Everything is O(1) here, too.
